ksmt-z3/src/main/kotlin/io/ksmt/solver/z3/ScopedFrame.kt

```diff
@@ -0,0 +1,129 @@
+package io.ksmt.solver.z3
+
+internal interface ScopedFrame<T> {
+    val currentScope: UInt
+    val currentFrame: T
+
+    fun flatten(collect: T.(T) -> Unit): T
+
+    /**
+     * find value [V] in frame [T], and return it or null
+     */
+    fun <V> find(predicate: (T) -> V?): V?
+
+    fun push()
+    fun pop(n: UInt = 1u)
+}
+
+internal class ScopedArrayFrame<T>(
+    currentFrame: T,
+    private val createNewFrame: () -> T
```

Comment:
Since `AraryFrame` is always used with `Long2ObjectOpenHashMap` it is better to create specialized version and remove lambdas